### DISCUSSION TITLE
Formula used when glass type specified

### DISCUSSION THREAD

**Discussion by dibyendumajumdar** [mjhoptics/ray-optics#40] [Q&A]
Hi,

Does ray-optics use exact refractive index per wavelength given in the glass catalogue?

**Comment by mjhoptics** [comment]
If you use the rindex() method of the glass, you're getting the interpolated refractive index, obtained by plugging the wavelength into the interpolating polynomial, using the manufacturer's coefficients. There's another way to query the data in the catalog spreadsheet, and that returns the measured refractive index, for a small set of spectral lines. Here's an example:

`from rayoptics.environment import *`

`FKH2 = create_glass('J-FKH2,Hikari')`

`wvls = FKH2.catalog.nline_str.keys(); wvls`
`Out[4]: dict_keys(['t', 's', 'r', 'C', "C'", 'D', 'd', 'e', 'F', "F'", 'g', 'h', 'i'])`

`for w_str in wvls:`
`        n_line = FKH2.glass_item(FKH2.catalog.nline_str[w_str])`
`        n_intrp = FKH2.rindex(w_str)`
`        print(f'{w_str}: {n_line}, {n_intrp:9.6f}, {n_intrp-n_line:8.2g}')`
    
`t: 1.45028,  1.450277,   -3e-06`
`s: 1.45173,  1.451728, -2.3e-06`
`r: 1.453592,  1.453592, -1.6e-07`
`C: 1.454469,  1.454469,  1.5e-07`
`C': 1.454714,  1.454714,   -3e-07`
`D: 1.455955,  1.455955,  3.4e-07`
`d: 1.456,  1.456000,  5.4e-09`
`e: 1.457192,  1.457192, -1.2e-07`
`F: 1.45946,  1.459460, -4.2e-07`
`F': 1.45974,  1.459740,  4.3e-07`
`g: 1.462126,  1.462126,    3e-07`
`h: 1.464317,  1.464317, -1.9e-07`
`i: 1.468003,  1.468003, -2.2e-07`

The interpolating polynomial matches the measured results to well within the quoted measurement error.

**Comment by dibyendumajumdar** [comment]
Thank you - I asked because I noticed that when I started to use glass types in one example, the 'red' dots became more diffused - I should expect the d-line to match the raw refractive index data I think. I haven't yet checked what values I am getting but I can use above to validate

**Comment by dibyendumajumdar** [comment]
I checked the values . Although the index matches to a significant digits, the result has trailing digits:

```
from rayoptics.environment import *

gls = create_glass('J-BK7A','Hikari')
print(gls.rindex('d'))
```

Gives 

```
1.5168000094330314
```

The expected value is `1.5168`.

I am guessing that this discrepancy although small is causing the results for d-line to be worse than when they are when I just use the index figures.

**Reply by mjhoptics** [reply]
A difference of 1e-8 should be insignificant for any practical usage. The manufacturers tightest tolerance on index (for their highest quality/homogeneity glasses) is 3e-6, IIRC.

Just to be sure, when you changed the material to the catalog glass, you followed up with 
`opm.update_model()`
right?

Just for background, when manufacturing a high performance lens, designers will often do a "melt fit" where the measured indices from the actual melt of glass being used are put into the model and the airspaces readjusted to compensate for small differences between the catalog values (which are averaged over many melts) and the specific melt used. These differences are in the 1e-6 to 1e-5 range.

**Comment by dibyendumajumdar** [comment]
I have opened a PR for the lens above. As you will see when glass types are specified the spot diagrams show that the red dots have got more spread out compared to the first example.

https://github.com/mjhoptics/ray-optics-notebooks/pull/9

**Reply by mjhoptics** [reply]
Thanks for the examples. I won't be able to make any definitive conclusions without a version of "ground truth" to compare to. But I think it'll be worth examining the two models and identify in the analyses the effect of the different glass approximations used.

**Comment by dibyendumajumdar** [comment]
Btw it will good to know what pattern ray-optics uses for the spot diagram - and also for the location of the skew ray spot diagram on the image plane. Another lens where the results are somewhat unexpected is this one:

https://github.com/mjhoptics/ray-optics-notebooks/blob/master/Photo-Lens-Patents/Nikkor-120-300mm-f2.8-JP2020-177057-Example-1.ipynb

The skew ray plot is unexpectedly bad. I cannot be sure that the patent data is totally correct - but when I plot this in GNU Optical I don't see a similar result. Of course the differences could be caused by the pattern, number of rays, and also the location on the image plane.

**Reply by mjhoptics** [reply]
What would be helpful in this case to to supply surface by surface ray trace data for the on-axis marginal ray, to the central wavelength (d?) and for the blue and/or red. There's too much high level implementation detail to try to compare spot diagrams. The SpotDiagramFigure in rayoptics uses a rectangular grid. The newer interface (in rayoptics.raytr.analyses) allows you to pass in the ray grid generator into RayList for more control over the pupil sampling.

**Comment by dibyendumajumdar** [comment]
Okay - I don't have a way to compare the ray trace data yet - I will get back re this.

**Comment by dibyendumajumdar** [comment]
Hi, 
I wanted to check following:

a) How is the rindex for C and F calculated from the abbe number and nd.
b) How is the rindex of air calculated for different wavelength.
c) Is there a way for me to trace a ray if I know the starting direction of the ray - assuming source is located at infinity.

Thanks and Regards
Dibyendu

**Reply by mjhoptics** [reply]
Hi Dibyendu,

a) Because the abbe number and nd aren't sufficient alone to construct an index model, I have to make an additional assumption. I assume the glasses are on the "normal glass line" and attempt to construct a model using a quadratic Buchdahl model. What constitutes a "normal glass" in what I've done is a glass that lies along a best fit line thru glasses plotted against the 2 Buchdahl chromatic coordinates. The Buchdahl2 class in the opticalglass.buchdahl module is where this is implemented; there is a default glass line that can be overriden by including a linear model in addition to the nd and abbe number. Different programs use different approaches to fill in the extra data; CODE V uses a model that depends on 3 "normal" glasses, for instance.

b) Glass data provided by the manufacturers is almost always supplied relative to air at standard temperature and pressure. So the refractive index of air is 1 at all wavelengths in this case.

c) The ray trace code is written in several layers with the lowest level being the `trace_raw` function in raytr.raytrace module. The next layer is in the raytr.trace module, with the functions trace() and trace_base(). For what you're asking to do, I would use the trace() function in the trace module. For example, if you want to trace a ray from infinity into the lens at 25 units from the axis at 587nm, you would do:
`ray_data = trace(sm, np.array([0., 25.0, 0.]), np.array([0., 0., 1.]), 587.)`
`list_ray(ray_data)`
The 2nd argument is the starting ray coordinate at the object surface and the 3rd argument is the direction cosine vector. 

One thing to note is that it should be possible to override the refractive indices calculated for the model with values that you calculate yourself. The SequentialModel has an attribute, `rndx`, that is a list of refractive indices at each gap for each wavelength. You can do:
`sm.rndx`
at the prompt to see the current refractive indices for the model. If you want to override these values to test the ray trace, say with an array of indices called `my_rndxs`, I would try:
`opm.update_model()`
`sm.rndx = my_rndxs`
`ray_data = trace(sm, np.array([0., 25.0, 0.]), np.array([0., 0., 1.]), 587.)`
`list_ray(ray_data)`

Hope this helps.

**Reply by dibyendumajumdar** [reply]
Thank you for the detailed reply. Re ray tracing - for an object an infinity, I am not sure how to specify an object surface. A related question is re the angles to use - if I want to simulate a skew ray at 21 degree angle, how would I do that? I ask because I feel there are multiple ways to specify this,  and it seems to vary across systems.

**Reply by mjhoptics** [reply]
The object surface is defined as interface 0. The object distance is:
`sm.gaps[0].thi=1e10`
A large distance like 1e10 is considered an acceptable substitute for infinity.
To compute the starting ray height for a ray at 21 deg, compute:
`obj_angle = radians(21)`
`y_ht = -sm.gaps[0].thi * tan(obj_angle)`
`obj_pt = np.array([0, y_ht, 0])`
`obj_dir_cos = np.array([0, sin(obj_angle), cos(obj_angle)])`
`ray_data = trace(sm, obj_pt, obj_dir_cos, 587.)`

You are right that there are different ways of setting up a ray trace, even within a given program. The example here uses the low level api which gives the most control over the ray specification. The next level up is to specify rays in terms of relative pupil coordinates and relative field height/angle. All optical design software has this in some form, ray-optics included. The trace_base() function is the basic api for using a field point and a relative pupil coordinate. More complicated functionality includes specifying the vignetted pupil per field point.

**Reply by dibyendumajumdar** [reply]
Hi Mike that's great, thank you so much.
